libmscore/layout.cpp

```diff
@@ -3845,7 +3845,13 @@ void Score::layoutSystemElements(System* system, LayoutContext& lc)
                               f->layout();
                               if (f->addToSkyline()) {
                                     Note* n = f->note();
-                                    QRectF r = f->bbox().translated(f->pos() + n->pos() + n->chord()->pos() + s->pos() + s->measure()->pos());
+                                    QRectF r = f->bbox().translated(f->pos() + n->pos() + n->chord()->pos());
+                                    // segment shapes were regenerated for beamed notes
+                                    // fingering shape was lost since it was cleared above
+                                    // so add it back now
+                                    if (n->chord()->beam())
```

Comment:
Do you mean [these lines](https://github.com/musescore/MuseScore/blob/54af16245874f210ddc21da2a4404f8ca4a7ab50/libmscore/layout.cpp#L3754-L3776) when talking about clearing fingerings' shapes? If so, does that operation really depend on beams? I don't see any checks for beams before doing that layout reset, and the purpose of it seems to be just resetting fingerings layout to prevent their potentially incorrect addition to segment shapes and skylines (as they are not laid out yet). This doesn't look like something related to beams.

Reply:
Well, that's part of what I mean.  The layout for the fingerings themselves is cleared for all notes indeed, but they were included in the chord shape when the *segment* shapes were generated, which is what matters for slurs more than the skyline.  However, beam layout ends up regenerating the segment shapes, and at that point the fingerings are gone from the chord shape, so they don't get added to the segment shape.  Hence, my PR does exactly that - makes sure they get added to the segment shape upon layout, since the layout happens after segment shapes are generated.  it's the same approach sued for articulations, which also need to be laid out fairly late to do the correct things with respect to slurs.

Reply:
OK, I see now, those lines pointed by me seem to affect skylines creation only.

But does that mean that chord-based fingerings layout should be cleared before laying out beams? In one of my recent PRs I [proposed](https://github.com/musescore/MuseScore/pull/5118/commits/f15a43b99fca6888c62281403dd027cbebb12b1b) doing layout of beams before creating skylines in order to avoid stems being added to skylines with incorrect lengths. That eventually results in reordering of that fingerings layout clearing and beams layout. Should I correct that in my PR, and would that change affect this fix?

Reply:
The clearing in those lines is about removing them from the chord's shape specifically.  At that point they've already been added to the segment shape, but we need to keep them out of the skyline.  Otherwise, when we go to lay them out, they would see themselves already in the skyline because they were still in the chord shape from the previous layout.  It's kind of a hack, but it was needed because fingerings are unique in that they are considered part of the chord shape for the purpose of creating segment shapes, and yet they can't actually be laid out until after we have already used the chord shape to create segment shapes.  There were other hacks I could have used instead, but this is what I chose.  It could certainly stand to be revisited.  In fact, with my PR here, it might be possible to simply remove them from the chord shape, and then the clearing wouldn't be needed either.  Hmmm.

Anyhow, I don't know for sure if it matters if the clearing is done before beam layout or not.  What's important is it be done before the chord is added to the skyline.  If beam layout were moved earlier, this could be relevant if the beam layout actually accesses the chord *shape* in its calculations.  My guess is it doesn't, that it works mostly off the chord position and stem length.

FWIW, though, there are other issues with beam layout that make me wonder if we need to restructure things even more.  In particular, to get for cross-measure beaming to really work across system breaks.  You made some changes that helped the situation a lot back in the fall, but it still leaves us with bugs like https://musescore.org/en/node/290088, and we still don't what is really wanted here, which is to have the beam extend past the end of the beam group to show the continuation.

There is also #4784.  The ability to have accidentals (and more) tuck under notes of another voice was an important part of @wschweer 's early work on 3.0, and it's a shame to have lost this.  That PR here would restore that feature, but it doesn't solve a problem that was presumably there all along, that stem length isn't known yet for beamed notes.  Fixing that would require beam layout to be moved earlier still, but I think there would be a chicken * egg problem in that chord layout needs to know something about beams and beam layout need to know something about chords.

All of that is to say, I think you should do what we need to for beams first and then let me take another look at this PR.  Just check that you don't see anything strange with fingering on beamed notes, like them jumping up higher and higher with each layout (that's what the original clearing code was designed to prevent, and not just for beamed notes).